### ISSUE TITLE
Can not load .glTF file in three.js editor r82dev + GLTFLoader v2

### ISSUE THREAD

**Issue by cx20** [mrdoob/three.js#9866]
##### Description of the problem
1. Access to following URL
   https://cdn.rawgit.com/mrdoob/three.js/dev/editor/index.html
2. [File] - [Import] - [Duck.gltf] select Duck.gltf file on the desktop
   -> Can not load Duck.gltf.  Following error has been displayed on the console.

```
three.min.js:150 THREE.WebGLRenderer 82dev
Storage.js:75 [22:07:15] Saved state to IndexedDB. 1.94ms
Loader.js:22 Loading Duck.gltf (191 KB) 100%
GLTFLoader.js:77 Synchronous glTF object access is deprecated. Use the asynchronous 'callback' argument instead.get scene @ GLTFLoader.js:77(anonymous function) @ Loader.js:168FileReader (async)Loader.loadFile @ Loader.js:173(anonymous function) @ Menubar.File.js:45
GLTFLoader.js:79 Uncaught ReferenceError: scene is not defined
    at Object.get scene [as scene] (https://cdn.rawgit.com/mrdoob/three.js/dev/examples/js/loaders/GLTFLoader.js:79:12)
    at FileReader.<anonymous> (https://cdn.rawgit.com/mrdoob/three.js/dev/editor/js/Loader.js:168:14)get scene @ GLTFLoader.js:79(anonymous function) @ Loader.js:168FileReader (async)Loader.loadFile @ Loader.js:173(anonymous function) @ Menubar.File.js:45
GLTFLoader.js:59 GLTFLoader: 130ms
GLTFLoader.js:67 Uncaught (in promise) TypeError: callback is not a function
    at https://cdn.rawgit.com/mrdoob/three.js/dev/examples/js/loaders/GLTFLoader.js:67:4
    at GLTFParser.<anonymous> (https://cdn.rawgit.com/mrdoob/three.js/dev/examples/js/loaders/GLTFLoader.js:813:3)(anonymous function) @ GLTFLoader.js:67(anonymous function) @ GLTFLoader.js:813Promise.resolve (async)_each @ GLTFLoader.js:593(anonymous function) @ GLTFLoader.js:1710Promise.resolve (async)_each @ GLTFLoader.js:593(anonymous function) @ GLTFLoader.js:1517Promise.resolve (async)_each @ GLTFLoader.js:593(anonymous function) @ GLTFLoader.js:1246Promise.resolve (async)_each @ GLTFLoader.js:593(anonymous function) @ GLTFLoader.js:955
```

In previous version (r81 + GLTFLoader v1) it will be displayed correctly.
https://cdn.rawgit.com/mrdoob/three.js/master/editor/index.html
##### Three.js version
- [x] Dev
- [ ] r81
- [ ] ...
##### Browser
- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer
##### OS
- [x] All of them
- [ ] Windows
- [ ] Linux
- [ ] Android
- [ ] IOS
##### Hardware Requirements (graphics card, VR Device, ...)

Windows 10 + ThinkPad X201 Intel HD Graphics

**Comment by mrdoob** [comment]
Fixed!

Unfortunately, seems like all the `glTF-Embedded` sample files have custom shaders embedded and they don't render properly in the editor...

Neither the editor can yet load multiple files (Duck.bin, Duck.gltf, DuckCM.png) which is what `glTF-MaterialsCommon` sample files provide.

/cc @richtr

**Comment by cx20** [comment]
@mrdoob Thank you for fix.

The Duck came to be displayed. However, it is out of the Bounding Box. 
And I cannot move it if I drag a Bounding Box with a mouse.

![image](https://cloud.githubusercontent.com/assets/1833979/19484953/6c2a6d58-9594-11e6-979e-2070d7ac08b0.png)

I tried the following Duck.gltf.
https://github.com/KhronosGroup/glTF/blob/master/sampleModels/Duck/glTF-Embedded/Duck.gltf

**Comment by mrdoob** [comment]
Yeah, as I said, that's because it's using custom shaders... If you change the material of the object it should display properly.

**Comment by mrdoob** [comment]
@pjcozzi any plans on adding `glTF-Embedded-MaterialsCommon`? 😅

**Comment by cx20** [comment]
@mrdoob I'm sorry, I finally understood it.

Previous version (r81 + GLTFLoader v1) does not use custom shader.
Next version (r82dev + GLTFLoader v2) uses custom shader.

**Comment by pjcozzi** [comment]
> @pjcozzi any plans on adding glTF-Embedded-MaterialsCommon?

We have plans to update the sample models and this could certainty be part of it, @lexaknyazev.

**Comment by mrdoob** [comment]
@cx20 yes, I'm having the same issue.

**Comment by cx20** [comment]
@mrdoob I understood of custom shader problem. It is annoying problem...

http://jsdo.it/cx20/QwSL

**Comment by mrdoob** [comment]
@richtr Is there a way to pass the projection matrix to these custom shaders so they render properly?

**Comment by cx20** [comment]
@mrdoob @richtr

There was a problem with my test code.
http://jsdo.it/cx20/M7WC

```javascript
function render() {
    THREE.GLTFLoader.Shaders.update(scene, camera); // Add
    renderer.render(scene, camera);
}
```

Perhaps, also three.js editor, I think that there is a need to add a similar code.

**Comment by mrdoob** [comment]
Good to see what the solution is!
I'm still unsure embeding shaders inside the model file is a good idea though... For instance, the shading doesn't change in the test code if the duck rotates.

**Comment by webprofusion-chrisc** [comment]
Ah, this could be related to a problem I'm having. I naively loaded a .obj converted to .gltf then tried to texture it as normal. Looking at the .gltf it does indeed have embedded shaders and materials - I only want the mesh (I'm building a guitar configurator, so I apply materials etc dynamically). Hmm, wonder how if I can discard that from the loaded .gltf?

**Comment by mrdoob** [comment]
@webprofusion-chrisc how are you converting to .gltf?

**Comment by webprofusion-chrisc** [comment]
@mrdoob the original model is rhino 3D, I'm exporting .obj for part of the model (geometry only). I then use obj2gltf with default parameters to export the gltf. I note that this introduces default materials and shaders. 

What I'm trying to do is just load the mesh then apply my own materials, which works just using obj loader. With GLTFLoader I load the model, then have tried using either the gltfobject.scene or gltfobject.scene.children[0] as my Object3D, then continue from there as if I'd loaded it via OBJLoader etc. 

I'm wondering if I need to clone geometry or something to disassociate from the rest of GLTF as I feel it's trying to use the default shaders etc. Using .scene as the Object3D preserves the original scale and translation, using .scene.children[0] messes with the scale (making it 100 times the size). For either approach applying a texture(.map)  gives the model a single colour instead of stretching the texture over the model surface (so instead of looking like wood, it just looks brown).

I'm just not sure if how I'm using it is going against the actual design of how it's supposed to be used.

**Comment by mrdoob** [comment]
Maybe obj2gltf has options for not including its own shaders?

**Comment by webprofusion-chrisc** [comment]
@mrdoob conscious off pulling this issue off topic but yes it does have a few options, currently not one to discard shaders/materials. I was hoping to work around that by just using the data I need and ignoring extra info (the default shaders etc). Should I perhaps open a new GLTFLoader issue and ping the GLTFLoader authors?

**Comment by mrdoob** [comment]
One workaround would be to traverse the resulting `scene` and change the materials to your own.

**Comment by webprofusion-chrisc** [comment]
@mrdoob thanks, yes tried that - there must be some other issue because doing that does nothing, tried cloning the object, removing all GLTF Shaders etc. Just resorting to hacks now. Have actually temporarily given up as the app needs to be ready to demonstrate soon, so reverting to .obj. I think longer term I need to file a separate issue and get some good example code together.

**Comment by mrdoob** [comment]
Can you share your code?

**Comment by webprofusion-chrisc** [comment]
@mrdoob not a very useful amount of it no, I'll see if I can do a simple example instead.

**Comment by donmccurdy** [comment]
The original issue (RawShaderMaterial did not update correctly) has been fixed by #10851.

I think that re-assigning an object's material should 'just work' (maybe also unset `mesh.onBeforeRender`?) but let's open a new issue if there's a problem there. GLTFLoader had been keeping an internal reference to these shaders, but isn't anymore, so maybe that was the problem.

**Comment by cx20** [comment]
I confirmed that it was fixed.
